lib/storage/src/content_manager/migration.rs

```diff
@@ -0,0 +1,172 @@
+use collection::shards::shard::{PeerId, ShardId};
+use collection::shards::CollectionId;
+use segment::types::{WithPayloadInterface, WithVector};
+use tokio::sync::RwLock;
+
+use crate::content_manager::collections_ops::Collections;
+
+const MIGRATION_BATCH_SIZE: usize = 1000;
+const COLLECTION_INITIATION_TIMEOUT: Duration = Duration::from_secs(60);
+
+/// Handlers for migration data from one collection into another within single cluster
+
+/// Get a list of local shards, which can be used for migration
+///
+/// For each shard, it should present on local peer, it should be active, it should have maximal peer id.
+/// Selection of max peer id guarantees that only one shard will be migrated from one peer.
+async fn get_local_source_shards(
+    source: &Collection,
+    this_peer_id: PeerId,
+) -> CollectionResult<Vec<ShardId>> {
+    let collection_state = source.state().await;
+
+    let mut local_responsible_shards = Vec::new();
+
+    // Find max replica peer id for each shard
+    for (shard_id, shard_info) in collection_state.shards.iter() {
+        let responsible_shard_opt = shard_info
+            .replicas
+            .iter()
+            .filter(|(_, replica_state)| **replica_state == ReplicaState::Active)
+            .max_by_key(|(peer_id, _)| *peer_id)
+            .map(|(peer_id, _)| *peer_id);
+
+        let responsible_shard = match responsible_shard_opt {
+            None => {
+                return Err(CollectionError::service_error(format!(
+                    "No active replica for shard {}, collection initialization is cancelled",
+                    shard_id
+                )));
+            }
+            Some(responsible_shard) => responsible_shard,
+        };
+
+        if responsible_shard == this_peer_id {
+            local_responsible_shards.push(*shard_id);
+        }
+    }
+
+    Ok(local_responsible_shards)
+}
+
+fn handle_get_collection(collection: Option<&Collection>) -> CollectionResult<&Collection> {
+    match collection {
+        Some(collection) => Ok(collection),
+        None => Err(CollectionError::service_error(
+            "Collection is not found".to_string(),
+        )),
+    }
+}
+
+async fn migrate_shard(
+    collections: Arc<RwLock<Collections>>,
+    source_collection_name: &CollectionId,
+    target_collection_name: &CollectionId,
+    shard_id: ShardId,
+) -> CollectionResult<()> {
+    let mut offset = None;
+    let limit = MIGRATION_BATCH_SIZE;
+
+    loop {
+        let request = ScrollRequest {
+            offset,
+            limit: Some(limit),
+            filter: None,
+            with_payload: Some(WithPayloadInterface::Bool(true)),
+            with_vector: WithVector::Bool(true),
+        };
+
+        let collections_read = collections.read().await;
+
+        let source_collection =
+            handle_get_collection(collections_read.get(source_collection_name))?;
+        let _updates_guard = source_collection.lock_updates().await;
```

Comment:
If I understand this properly, this is not a transactional copy of the collection as it was at a given point in time.
As we progress through the `offset`, we will fetch the latest values for each point.
Update occurring on already fetched ids will be missed.

Am I understanding this right?

Reply:
Yes, we do not forward updates into the new collection automatically. But the lock allows the user to start pushing updates into both collections from the external source, and, therefore, switch from one collection to another transparently.

Reply:
Thanks for the clarification, it could be in the docs as well :memo: